src/main/java/org/truffleruby/core/time/RubyDateFormatter.java

```diff
@@ -544,6 +568,168 @@ public static RopeBuilder formatToRopeBuilder(List<Token> compiledPattern, Zoned
+                        appendRope = RopeConstants.paddedNumber(dt.getMonthValue());
+                        break;
+                    case FORMAT_SECONDS:
+                        appendRope = RopeConstants.paddedNumber(dt.getSecond());
+                        break;
+
+                    case FORMAT_YEAR_LONG: {
+                        final int value = dt.getYear();
+
+                        if (value < 1000 || value > 9999) {
+                            CompilerDirectives.transferToInterpreterAndInvalidate();
+                            return formatToRopeBuilder(
+                                    compiledPattern,
+                                    dt,
+                                    zone,
+                                    context,
+                                    language,
+                                    currentNode,
+                                    errnoErrorNode).toRope();
+                        }
+
+                        appendRope = new LazyIntRope(value, UTF8Encoding.INSTANCE, 4);
+                    }
+                        break;
+
+                    case FORMAT_NANOSEC: {
+                        final int nano = dt.getNano();
+                        assert nano >= 0;
+                        assert nano < 1000000000;
+
+                        final int digits;
+
+                        if (nano < 10) {
+                            digits = 1;
+                        } else if (nano < 100) {
+                            digits = 2;
+                        } else if (nano < 1000) {
+                            digits = 3;
+                        } else if (nano < 10000) {
+                            digits = 4;
+                        } else if (nano < 100000) {
+                            digits = 5;
+                        } else if (nano < 1000000) {
+                            digits = 6;
+                        } else if (nano < 10000000) {
+                            digits = 7;
+                        } else if (nano < 100000000) {
+                            digits = 8;
+                        } else {
+                            digits = 9;
+                        }
+
+                        final LazyIntRope nanoRope = new LazyIntRope(nano, UTF8Encoding.INSTANCE, digits);
+
+                        final int padding = 6 - digits;
+
+                        if (padding == 0) {
+                            appendRope = nanoRope;
+                        } else if (padding < 0) {
+                            appendRope = new SubstringRope(UTF8Encoding.INSTANCE, nanoRope, 0, 6, 6, CodeRange.CR_7BIT);
+                        } else {
+                            appendRope = new ConcatRope(
+                                    nanoRope,
+                                    RopeConstants.paddingZeros(padding),
+                                    UTF8Encoding.INSTANCE,
+                                    CodeRange.CR_7BIT);
+                        }
```

Comment:
The logic for nanos feels like it deserves to be a node, then it would be easier to profile and to use ConcatNode & SubstringNode. Building those directly feels quite fragile and potentially suboptimal (e.g., doing a concat with an empty Rope, etc).